src/3d/castletransform_physics.inc

```diff
@@ -613,14 +603,42 @@ end;
 
 procedure TRigidBody.SetLinearVelocity(const LVelocity: TVector3);
 begin
-  FKraftBody.LinearVelocity := VectorToKraft(ZeroLockedComponents(LVelocity, FLockTranslation));
-  if not LVelocity.IsPerfectlyZero then
-    FKraftBody.SetToAwake;
+  if FKraftBody <> nil then
```

Comment:
The setter should *always* set CGE private field `FLinearVelocity` (regardless if FKraftBody <> nil).

Likewise the getter (TRigidBody.GetLinearVelocity below) should update the CGE private field `FLinearVelocity` (only if FKraftBody <> nil), and then always return `FLinearVelocity` (regardless if FKraftBody <> nil).

Same note for TRigidBody.SetAngularVelocity, TRigidBody.GetAngularVelocity below.

Reason:

My approach seems more limiting, but is more intuitive, and it is consistent with how TCastleTransform.Translation/Rotation behave. The goal is that these sequences of code should result in the 100% same state:

```
// doing SetLinearVelocity when FKraftBody is nil
RigidBody.LinearVelocity := Vector3(1, 2, 3);
MyTransform.RigidBody := RigidBody;
```

and

```
MyTransform.RigidBody := RigidBody;
// doing SetLinearVelocity when FKraftBody <> nil
RigidBody.LinearVelocity := Vector3(1, 2, 3);
```

These just swap 2 lines, and it is intuitive that they should achieve 100% the same state. A deserialization system (once we made RigidBody and it's properties published) could also execute these 2 lines in any order.

Right now, in your code, there is a subtle difference between these two sequences: if you set `Transform.RigidBody := nil`, then in one case `RigidBody.LinearVelocity` will switch back to zero (since FLinearVelocity was not modified). We want to avoid it. In general, doing `Transform.RigidBody := nil` should keep the last seen state of `RigidBody.LinearVelocity`.

Reply:
> Right now, in your code, there is a subtle difference between these two sequences: if you set Transform.RigidBody := nil, then in one case RigidBody.LinearVelocity will switch back to zero (since FLinearVelocity was not modified). We want to avoid it. In general, doing Transform.RigidBody := nil should keep the last seen state of RigidBody.LinearVelocity.

Yes it will work like that (there will be no difference when you set velocity before assigning rigid body or after assigning like in ). I just not finished it yesterday and pushed changes habitually. Sorry about this. I wanted add this values (linear/angular velocity and many others) synchronization code on `DeinitializeTransform`. I think this should work.

I agree with you that setters should update FLinearVelocity, and I will change that. But I have some doubts about getters. In setters we will set that because of serialization (I must remember about that!). Setting `FLinearVelocity` in getter will not guarantee that value will be correct and additionally can lower performance (for example when in user code someone will get linear velocity on each frame).  I do not know how much, but I preferred not to do something that you do not have to do.

I think better solution will be add function which synchronizes all the physics values like linear velocity. And run it before serialization. The same function will be called in `DeinitializeTransform`.
Then we will have certainty that all values are valid. We can name it `SynchronizePhysics` or something like that.  What do you think about that?

Reply:
I added `SynchronizeFromKraft`.  And after that I'm not sure. Maybe it would be better to set private CGE fields in the getter as you said to do not duplicate code and just call getters in `SynchronizeFromKraft`. On the one hand the code will be in one place and not duplicated but on the other hand I feel that the code will not be so optimal when we do that. Please check current implementation and say what do you think about that.